Support host.docker.internal in dockerd on Linux

arkodg opened the pull request:
- addresses https://github.com/docker/for-linux/issues/264 Support host.docker.internal DNS name to host
- replaces, closes https://github.com/docker/libnetwork/pull/2348 special DNS record for host.docker.internal + gateway.docker.internal

Other issues (linking for discoverability)

- addresses https://github.com/docker/for-win/issues/5167 Unable to connect to WSL2 host IP after update to 2.1.6.0
- addresses https://github.com/docker/for-mac/issues/2705 They're close, but "docker.for.mac.host.*" and "docker.for.win.*" should be bette
- addresses https://github.com/moby/moby/issues/36625 They're close, but "docker.for.mac.host.*" and "docker.for.win.*" should be better
- addresses https://github.com/docker/for-win/issues/1855 They're close, but "docker.for.mac.host.*" and "docker.for.win.*" should be better docker/for-win#1855
- addresses https://github.com/jtreminio/dashtainer/issues/2 Xdebug on Linux needs a programmatic value for remote_host
- addresses https://github.com/bscheshirwork/docker-yii2-app-advanced/issues/1 Docker image bscheshir/php:7.2.5-fpm-alpine-4yii2-xdebug cannot connect to the host machine for XDebug
- addresses https://github.com/drud/ddev/issues/843 v0.18.0 xdebug instructions don't work on linux due to no "host.docker.internal" hostname
- addresses https://github.com/jc21/nginx-proxy-manager/issues/259 host.docker.internal is not resolved by nginx
- addresses https://github.com/cypress-io/cypress-docker-images/issues/183 http://host.docker.internal Does not work
- addresses https://github.com/Decathlon/ara/issues/222 local installation does not work on Ubuntu (with Docker and docker compose)
- addresses https://github.com/davidfowl/Micronetes/issues/22 Support for container networks / multicontainer workloads

**- What I did**
This PR allows containers to connect to Linux hosts
    by appending a special string "host-gateway" to --add-host
    e.g. "--add-host=host.docker.internal:host-gateway" which adds
    host.docker.internal DNS entry in /etc/hosts and maps it to host-gateway-ip
    This PR also add a daemon flag call host-gateway-ip which defaults to
    the default bridge IP
    Docker Desktop will need to set this field to the Host Proxy IP
    so DNS requests for host.docker.internal can be routed to VPNkit

**- How to verify it**
```
dockerd &
docker run -it --add-host=host.docker.internal:host-gateway alpine cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.18.0.1	host.docker.internal
172.18.0.2	9f74a81028b7
```

```
dockerd --host-gateway-ip=1.2.3.4 &
docker run -it --add-host=host.docker.internal:host-gateway alpine cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
1.2.3.4	host.docker.internal
172.18.0.2	a5f53cc81d24
```

tao12345666333 commented:
Looks Good.  But I don't really understand the intent of this feature.

Just to save the IP information of docker0?  Is it the same effect when using environment variable records?

If I understand wrong, please tell me.

arkodg commented:
@tao12345666333 updated the PR with a link to the issue

arkodg commented:
cc @justincormack @mavenugo @thaJeztah @djs55

Dwood15 commented:
Yes, please :eyes:

djs55 commented:
@arkodg looks good to me. So on Docker Desktop we would add `--dns-resolve-docker-host=false` to the command line and this will disable the feature so that queries for `host.docker.internal` (and `gateway`) will be resolved by our host software as before.

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -118,6 +118,24 @@ func (daemon *Daemon) buildSandboxOptions(container *container.Container) ([]lib
+			}
+			if len(gateway) > 0 {
+				sboxOptions = append(sboxOptions, libnetwork.OptionExtraHost(network.DockerHostInternalDNSName, gateway))
+				sboxOptions = append(sboxOptions, libnetwork.OptionExtraHost(network.DockerGatewayInternalDNSName, gateway))
```
Actually; I see you mention the reverse;

> This PR allows containers to connect to Linux hosts by setting a Daemon flag called dns-resolve-docker-host which adds a host.docker.internal entry in /etc/hosts and maps it to docker0's IP

`/etc/hosts` is only used on the default (bridge) network, but isn't used for DNS resolution on other networks (well, I guess it's used, but it would be different from resolution of other entries we set in the embedded DNS)?


edit: hm.. so we actually _do_ use it for the custom hosts (`--add-host`), so this would be consistent. Would it be better to use the embedded DNS?

```bash
docker network create foo 

docker run --rm --network=foo --add-host foo.example.com:123.123.123.1 busybox cat /etc/hosts

127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
123.123.123.1	foo.example.com
172.19.0.2	de9b9822b00f
```

cpuguy83 replied:
What happens if this is, for example, a macvlan interface? Wouldn't the gateway be different (ie. not necessarily the "host" ip) in such a case?

thaJeztah commented on cmd/dockerd/config.go:
```diff
@@ -64,6 +64,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) error {
 	flags.Var(opts.NewListOptsRef(&conf.DNS, opts.ValidateIPAddress), "dns", "DNS server to use")
 	flags.Var(opts.NewNamedListOptsRef("dns-opts", &conf.DNSOptions, nil), "dns-opt", "DNS options to use")
 	flags.Var(opts.NewListOptsRef(&conf.DNSSearch, opts.ValidateDNSSearch), "dns-search", "DNS search domains to use")
+	flags.BoolVar(&conf.DNSResolveDockerHost, "dns-resolve-docker-host", true, "dockerd will resolve the host.docker.internal DNS name if set to true")
```
Not a big fan of making this a boolean. Perhaps we could make this flag (that can be specified multiple times) to add custom entries to the resolver (either the embedded DNS or (see my other comment) entries in `/etc/hosts`, depending on the situation).

For this particular case, we may need a "special" value for the gateway address (which I think is dynamically set), so that users can (haven't thought of the name of the flag yet);

```bash
--custom-dns-entry host.docker.internal=gateway \
--custom-dns-entry gateway.docker.internal= gateway
```

With that (perhaps) Docker Desktop could also use that option?

scottgeary replied:
This would also make it compatible using Compose's `extra_hosts:` field.

thaJeztah replied:
that option is per container/service, not at the daemon level

arkodg replied:
@thaJeztah for this case, the community prefers some `automatic`

arkodg replied:
also if we use embedded DNS it won't support this DNS name for containers attached to the default `bridge`/`docker0` network

thaJeztah replied:
> @thaJeztah for this case, the community prefers some automagic

While I understand that some magic could be nice to have, I think the reason we originally didn't add it on Linux was specifically because hard-coding could cause issues in some situations;

For reference, the `host.docker.internal` was added on Docker Desktop to get parity with Linux (see https://github.com/moby/moby/issues/22753: "No docker0 on docker for mac?"). 

Related discussions use-cases / issues were

- https://github.com/moby/moby/issues/1143 Document how to connect to Docker host from container
- https://github.com/moby/moby/issues/23177 Add dockerhost as an entry in /etc/hosts in all containers
- (slightly related) https://github.com/moby/moby/issues/15086 "Document how to get real remote client ip for service running in container"


I _think_ that;

- For the first issue; on Docker Desktop, docker has (almost) full control over how networking is set-up, so implementing `host.docker.internal` could be done because of that, whereas on Linux, networking configuration can widely differ, so hard-coding values may cause issues in some setups
- For the other issues, expectations "how to connect to the host" widely varied;
    - some people were looking at connecting to a (non-containerised) service running on the same host
    - some people were looking at "how to connect to a service running on the same host" (either containerised or non-containerised) on a "interface", and how to get that IP-address?
    - "public interface" / IP-address is ambiguous in setups where the docker host is behind a proxy (and the docker host itself does not know the public address

To get "magic", docker could still decide to set these options in the default configuration (systemd unit?), but hard-coding could potentially (see above) limit the usefulness if users cannot customise these settings, and if the hard-coded values don't work.

arkodg replied:
@thaJeztah the hard coding can always be removed by setting `dns-resolve-docker-host=false` if someone wants to be change this

thaJeztah replied:
But for the situations mentioned above, there would be no way to change `host.docker.internal=<hard-coded-ip>` to `host.docker.internal=<correct IP>` to be included in all containers.

arkodg replied:
`dns-resolve-docker-host=false` (In Daemon config) && `--add-host host.docker.internal=<some-IP>` would address your concern

There is no ideal answer (I have raised the same questions), but apparently many developers want parity across OSs with the same Compose file and this PR attempts to do that

thaJeztah replied:
it would have to be set on every container, every service, so not the same as this feature implements

thaJeztah commented:
Seeing some failures;

s390x and powerpc integration tests (looks legit); https://ci.docker.com/public/job/moby/job/PR-40007/1/execution/node/217/log/

```
00:17:27.053  === RUN   TestLinksEtcHostsContentMatch
00:17:27.943  --- FAIL: TestLinksEtcHostsContentMatch (0.65s)
00:17:27.943      links_linux_test.go:33: assertion failed: 
00:17:27.943          --- ←
00:17:27.943          +++ →
00:17:27.943          @@ -5,4 +5,5 @@
00:17:27.943           ff02::1	ip6-allnodes
00:17:27.943           ff02::2	ip6-allrouters
00:17:27.943          -172.17.0.2	6d642f0c99aa
00:17:27.943          +172.18.0.1	host.docker.internal
00:17:27.943          +172.18.0.1	gateway.docker.internal
00:17:27.943           


....


00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon (1.97s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon (1.97s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=false/container_without_restart_policy/kill-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false] : [d3521c6b8cec2] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon d3521c6b8cec2 is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon (2.14s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon (2.14s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false --live-restore] : [d88a76ccc6bfb] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon d88a76ccc6bfb is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon (1.93s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon (1.93s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false --live-restore] : [d2ec219ca71f4] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon d2ec219ca71f4 is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon (1.95s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon (1.95s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/stop-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false --live-restore] : [de1b3c56d6a31] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon de1b3c56d6a31 is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon (1.92s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon (1.92s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false] : [df42d55b97be8] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon df42d55b97be8 is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/kill-daemon (1.94s)
00:19:01.889  === PAUSE TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/kill-daemon
00:19:01.889  === CONT  TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/kill-daemon
00:19:01.889      --- FAIL: TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/kill-daemon (1.94s)
00:19:01.889          restart_test.go:67: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDaemonRestartKillContainers/live-restore=true/container_without_restart_policy/kill-daemon"
00:19:01.889          restart_test.go:89: failed to start daemon with arguments [--iptables=false --live-restore] : [d392e42874177] Daemon exited during startup: exit status 2
00:19:01.889          panic.go:563: Daemon d392e42874177 is not started
00:19:01.889  
00:19:01.889  === FAIL: s390x.integration.container TestDaemonRestartKillContainers (0.00s)
```



amd64 integration-cli tests: https://ci.docker.com/public/job/moby/job/PR-40007/1/execution/node/196/log/

```
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestCleanupMountsAfterDaemonCrash (2.46s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestCleanupMountsAfterDaemonCrash (2.46s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestCleanupMountsAfterDaemonCrash"
00:58:54.789          docker_cli_daemon_test.go:2016: failed to start daemon with arguments [--live-restore] : [dc5633eb3befd] Daemon exited during startup: exit status 2
00:58:54.789          check_test.go:309: Daemon dc5633eb3befd is not started
00:58:54.789  
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonRestartRestoreBridgeNetwork (2.14s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestDaemonRestartRestoreBridgeNetwork (2.14s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestDaemonRestartRestoreBridgeNetwork"
00:58:54.789          docker_cli_network_unix_test.go:1644: failed to start daemon with arguments [--live-restore] : [d1dab7dde500b] Daemon exited during startup: exit status 2
00:58:54.789          panic.go:563: Daemon d1dab7dde500b is not started
00:58:54.789          check_test.go:309: Daemon d1dab7dde500b is not started
00:58:54.789  
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonRestartWithUnpausedRunningContainer (2.12s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestDaemonRestartWithUnpausedRunningContainer (2.12s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestDaemonRestartWithUnpausedRunningContainer"
00:58:54.789          docker_cli_daemon_test.go:2081: failed to start daemon with arguments [--live-restore] : [d2a14381f4c28] Daemon exited during startup: exit status 2
00:58:54.789          panic.go:563: Daemon d2a14381f4c28 is not started
00:58:54.789          check_test.go:309: Daemon d2a14381f4c28 is not started
00:58:54.789  
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.28s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.28s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore"
00:58:54.789          docker_cli_daemon_test.go:2720: failed to start daemon with arguments [--live-restore] : [d245f49a55601] Daemon exited during startup: exit status 2
00:58:54.789          check_test.go:309: Daemon d245f49a55601 is not started
00:58:54.789  
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestRemoveContainerAfterLiveRestore (2.69s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestRemoveContainerAfterLiveRestore (2.69s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestRemoveContainerAfterLiveRestore"
00:58:54.789          docker_cli_daemon_test.go:2739: failed to start daemon with arguments [--live-restore --storage-driver overlay] : [d8664dd6e9cf8] Daemon exited during startup: exit status 2
00:58:54.789          check_test.go:309: Daemon d8664dd6e9cf8 is not started
00:58:54.789  
00:58:54.789  === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestRestartPolicyWithLiveRestore (1.94s)
00:58:54.789      --- FAIL: TestDockerDaemonSuite/TestRestartPolicyWithLiveRestore (1.94s)
00:58:54.789          daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestRestartPolicyWithLiveRestore"
00:58:54.789          docker_cli_daemon_test.go:2782: failed to start daemon with arguments [--live-restore] : [d21d61da22ae8] Daemon exited during startup: exit status 2
00:58:54.789          check_test.go:309: Daemon d21d61da22ae8 is not started
```

Looks like there's a NPE: 
[docker.log](https://github.com/moby/moby/files/3672547/docker.log)



```
time="2019-09-29T08:12:20.166341466Z" level=debug msg="container mounted via layerStore: &{/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestCleanupMountsAfterDaemonCrash/dc5633eb3befd/root/overlay2/b27cbc0c172353a0deb8cde0698da92bdaefbb7a374ac3c64a2e69cf1d3c2173/merged 0x556347a54840 0x556347a54840}"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x55634536af90]
```

arkodg commented:
CI failed due to https://github.com/moby/moby/pull/39966#issuecomment-536789150

cpuguy83 commented:
There seems to be a fair number of edge cases this has a lot more edge cases than are accounted for here.
For example, what happens on a macvlan network where the "gateway" is not necessarily (not even likely) the host IP.

In Docker4Mac there is a very specific and controlled network stack to get at the "host" IP.

arkodg commented:
@cpuguy83 the above PR attempts to route `host-docker-internal` to the Host, and so the `default ` network's IP is a convenient way to retrieve that, AFAIK `default` network is always based off the `bridge` driver and not `macvlan` . In case the default bridge is disabled, these entries will not be added

cpuguy83 commented:
Ok, SGTM.

tiborvass commented:
This should be on `docker run` and not a daemon flag, because non-portable properties are in HostConfig and we should not add more non-portable properties to the daemon. We already have `--add-host` and the only problem is the IP is not portable. I like @thaJeztah's approach in https://github.com/moby/moby/pull/40007#discussion_r329662955 for add-host.

For swarmkit consideraitons: this would be opaque to the managers and they would pass it down to workers and each worker would resolve it to its own IP.

tiborvass commented:
Here's a conclusion of my IRL talk with @arkodg:
- we should not break existing Desktop users
- instead of adding yet another knob to the daemon that makes all applications running on it less portable, users should be encouraged to opt-in to non-portability per application.

Suggested implementation:
- `add-host` accepts as an IP, a magic `gateway` string (open to suggestions for naming) that, by default is replaced by the bridge IP (as configued by `dockerd --bip`)
- a new `--hosts-gateway-default-ip` flag (again: i'm sure we can find a better name) is added to `dockerd` which allows docker desktop to configure the IP that the magic `gateway` string should be replaced to.
- Docs are updated to deprecate (but not remove) the automatically added hosts in Desktop, and instead encourage users to add manually the magic host if needed.

As someone pointed it out, this would work automatically with compose using the `gateway` string in  `extra_hosts`.

thaJeztah commented:
I think in addition to the above, we could add a configuration on the CLI that allows setting these as a default, similar to what we did with proxy configuration; https://github.com/docker/cli/pull/93

arkodg commented:
@tibor
so what you are suggesting is a per container opt in strategy rather than a daemon level option since this usage is not the common-case
Now desktop will need to spawn `dockerd` with `--hosts-gateway-default-ip` `host-proxy-ip`.

Because in `dockerd` the code would look like
if `gateway` in `add-hosts`
get IP from `hosts-gateway-default-ip`.  (Desktop case)
if `hosts-gateway-default-ip` == nil
get IP from default bridge                            (dockerd running on host OS case)

wdyt @djs55

tiborvass commented:
@arkodg Almost: there is no conditional logic for getting the gateway IP: it's always from the `hosts-gateway-default-ip` config. That config defaults to the bridge IP upon daemon start.

djs55 commented:
This sounds reasonable to me. Adding an extra argument to `dockerd` is fine and if the per-container opt-in looks like `docker run --add-host gateway` then that's ok from my PoV.

Commit:
Support host.docker.internal in dockerd on Linux

Docker Desktop (on MAC and Windows hosts) allows containers
running inside a Linux VM to connect to the host using
the host.docker.internal DNS name, which is implemented by
VPNkit (DNS proxy on the host)

This PR allows containers to connect to Linux hosts
by appending a special string "host-gateway" to --add-host
e.g. "--add-host=host.docker.internal:host-gateway" which adds
host.docker.internal DNS entry in /etc/hosts and maps it to host-gateway-ip

This PR also add a daemon flag call host-gateway-ip which defaults to
the default bridge IP
Docker Desktop will need to set this field to the Host Proxy IP
so DNS requests for host.docker.internal can be routed to VPNkit

Addresses: https://github.com/docker/for-linux/issues/264

Signed-off-by: Arko Dasgupta <arko.dasgupta@docker.com>

arkodg commented:
updated the PR , would appreciate suggestions for a better name, picked `host-gateway` as a placeholder

```
docker run -it --add-host=host-gateway alpine cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.18.0.1	host.docker.internal
172.18.0.1	gateway.docker.internal
172.18.0.2	e25baec2305e
```

arkodg commented:
@tiborvass PTAL

tiborvass commented on cmd/dockerd/config.go:
```diff
@@ -64,6 +64,8 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) error {
 	flags.Var(opts.NewListOptsRef(&conf.DNS, opts.ValidateIPAddress), "dns", "DNS server to use")
 	flags.Var(opts.NewNamedListOptsRef("dns-opts", &conf.DNSOptions, nil), "dns-opt", "DNS options to use")
 	flags.Var(opts.NewListOptsRef(&conf.DNSSearch, opts.ValidateDNSSearch), "dns-search", "DNS search domains to use")
+	flags.StringVar(&conf.HostGatewayIP, "host-gateway-ip", "", "dockerd will resolve the IP Address to this IP "+
```
maybe something like `IP address that the special 'host-gateway' string in --add-host resolves to`

tiborvass replied:
Maybe use `opts.NewIPOpt` to have the flags package validate the IP format early on.

arkodg replied:
ack

tiborvass commented on daemon/container_operations.go:
```diff
@@ -115,6 +115,16 @@ func (daemon *Daemon) buildSandboxOptions(container *container.Container) ([]lib
+			if gateway := daemon.configStore.HostGatewayIP; net.ParseIP(gateway) != nil {
+				parts[1] = gateway
+			} else {
+				logrus.Warnf("HostGatewayIP value %s is invalid", gateway)
```
This should never happen if the config is validated on startup and HostGatewayIP is already a `net.IP`

arkodg replied:
there can be a case where default bridge is disabled and the HostGatewayIP is not set, in that case might be best to bail out ?

tiborvass commented:
Apart from minor nits, I'm fine with the PR. I don't have opinions on the names, so I'm fine with `host-gateway` magic string and `--host-gateway-ip` flag. If @djs55 @thaJeztah @tonistiigi or more opinionated people dislike the current names, speak up now.

thaJeztah reviewed (approved):
LGTM

one question; are `host.docker.internal` and `gateway.docker.internal` always expected to have the same "magic" value, or would there be situations where we'd wanted different values for both?

arkodg commented:
I'm not sure what the true difference between` gateway.docker.internal` and `host.docker.internal` is ? @djs55  might now

Sidenote - We need a similar fix in `docker-cli` in `opts/hosts.go`  once this PR is approved

perlun commented:
@cpuguy83 Any chance you could give this a look?

Dwood15 commented:
Anyone mind reviewing this? Don't want it to get lost.

djs55 commented:
@thaJeztah For reference on current docker desktop `host.docker.internal` is the magic host IP for connecting to services running on the host, which I think is the one users care about. They can do things like run native mac postgres from homebrew and connect to it from their container. It always sounds a bit strange, but I think it's seen as a way to incrementally containerise a dev setup: start with one component in a container, use the old on-host stuff for the rest, and slowly migrate to a fully containerised setup over time, rather than needing to do a "big bang" switchover.

We have a few other names in our DNS for other internal purposes; `gateway.docker.internal` is one of those: it's the IP of the default gateway in `vpnkit`. We also have `vm.docker.internal` which is the IP assigned to the VM -- we use this name in our k8s certs. I don't think users care about this internal desktop-specific plumbing though.

rakitadragan commented:
Is it guaranteed that a host IP will be the same in case that a host is rebooted?

AkihiroSuda commented:
The code looks good, but can we have an integration test?

AkihiroSuda reviewed (approved):
Thanks

AkihiroSuda commented on integration/internal/container/ops.go:
```diff
@@ -180,3 +180,11 @@ func WithCgroupnsMode(mode string) func(*TestContainerConfig) {
 	}
 }
+
+// WithExtraHosts sets the user defined IP:Host mappings in the container's
```
Golint failing on this typo

thaJeztah reviewed (approved):
missed that you updated; I see that `gateway.docker.internal` was removed 👍 

LGTM (pending the linting issue)

thaJeztah reviewed (approved):
still LGTM

moving to `merge`

romansh commented:
I'm a newbie in docker and a little confused. Can anyone explain is this going to be built into docker-for-linux and when or this is just a internal moby thing?

thaJeztah commented:
This will be in Docker for Linux (and Docker Desktop, which runs the Linux daemon, although inside a lightweight VM).

Difference between this and the current implementation on Docker Desktop is that;

- the current Docker Desktop implementation is in a part of the code-base that's proprietary (i.e., part of how Docker Desktop is configured internally)
- this code _could_ be used by the Docker Desktop team in future as well (to be discussed)
- this PR does not set up the "magic" `host.docker.internal` _automatically_ on every container, but it can be used to run a container that needs this host by adding `docker run --add-host host.docker.internal:host-gateway`
- (to be discussed); setting that "magic" domain automatically on containers that are started _could_ be implemented by adding an option for this in the `~/.docker/config.json` CLI configuration file.

thaJeztah commented:
All green now; merging this 👍

thaJeztah commented:
Opened https://github.com/docker/cli/issues/2290 to discuss options for the CLI

schildbach commented:
Question: will this `--add-host` option be available within docker-compose, too?

thaJeztah commented:
In a compose file, the corresponding `extra_hosts` option can be used; https://docs.docker.com/compose/compose-file/#extra_hosts

It's possible though that validation rules in docker-compose will have to be relaxed (if it's validating for valid IP-addresses to be specified, and not taking the "magic" value into account); similar to https://github.com/docker/cli/pull/2293

mcarleio commented:
does this pull request solve the problem to reference a local running proxy?
When I set ~/.docker/config.json to
```
{
 "proxies":
 {
   "default":
   {
     "httpProxy": "http://localhost:3128",
     "httpsProxy": "http://127.0.0.1:3128"
   }
 }
}
```
and then start `docker build .` (which fetches anything from http/https, e.g.`apt-get update`) then at the moment I will get something like:
`Could not connect to localhost:3128 (127.0.0.1). - connect (111: Connection refused)`

The only working workaround at the moment, is to use the machines hostname, so instead of `"httpProxy": "http://localhost:3128"` i am using `"httpProxy": "http://<MY_MACHINES_HOSTNAME>:3128"`.

Is this pull request allowing me to use `"httpProxy": "http://host.docker.internal:3128"` or something?

jeffrson commented:
Unfortunately, changes to /etc/hosts will not be usable in cases where a real resolver is needed - seen for example in nginx when there are variables in proxy_pass.

Currently, it's already possible to use something like "--add-host=host.docker.internal:172.17.0.1", which indeed adds a line to /etc/hosts. So if you have "proxy_pass host.docker.internal:8000" inside nginx config it works. However, if there's "proxy_pass host.docker.internal:$port" or any other usage of variables (regex comes to mind) it needs a resolver config entry, which (if 127.0.0.11 will be used) doesn't make use of /etc/hosts.

In Docker for Windows, btw, this works out of the box, without any "--add-host". Actually I don't understand, why there should be "--add-host" necessary on Linux for such a fundamental setting.

pleymor commented:
The only thing working for me was:

```bash
docker run --add-host host.docker.internal:$(dig +short myip.opendns.com @resolver1.opendns.com) my-super-image:1.0.0
```

Sebi2020 commented:
Why do I have to use an extra flag on linux while it works out of the box on windows and mac?! Inconsistencies over inconsistencies...

thaJeztah commented:
The `host.docker.internal` domain was added to Docker Desktop, to work around limitations of Docker Desktop because it's running inside a VM (which didn't apply to Linux), related to https://github.com/moby/moby/issues/1143.

On Docker Desktop, the `host.docker.internal` domain is provided through a DNS that's set up by Docker Desktop inside the VM. On Linux, Docker does not control the host's DNS, so it's not able to provide the same (but you could set up a DNS on your host to provide resolution for it as an alternative).

aradalvand commented:
This doesn't seem to be documented in the official documentation.
But why? It's a pretty important feature.

thaJeztah commented:
@aradalvand You're right; I see I added a `docs-revisit` label at the time (to indicate documentation changes are still needed after merging), but I guess that never was worked on. I opened https://github.com/docker/docs/issues/17597 for tracking

ongiant commented:
> The `host.docker.internal` domain was added to Docker Desktop, to work around limitations of Docker Desktop because it's running inside a VM (which didn't apply to Linux), related to #1143.
> 
> On Docker Desktop, the `host.docker.internal` domain is provided through a DNS that's set up by Docker Desktop inside the VM. On Linux, Docker does not control the host's DNS, so it's not able to provide the same (but you could set up a DNS on your host to provide resolution for it as an alternative).

It doesn't work even though I have added the `127.0.0.1  host.docker.internal` to my host machine `/etc/hosts`. My questions is written [here](https://stackoverflow.com/questions/77846380/how-can-i-leverage-host-docker-internal-for-simultaneous-use-in-both-dockers-in).

simonmysun commented:
Hi,

I wonder if it is possible to resolve `host.docker.internal` to an IPv6 address. My use case is to forward traffic from a IPv6 only interface to the host. It's better to find a IPv6 address than add a 6to4 converting layer, considering IPv4 will be phased out sooner or later. 

Thx!

thaJeztah commented:
@simonmysun will be coming in docker v28.0; see this PR;


- https://github.com/moby/moby/pull/48807

simonmysun commented:
Thanks! @thaJeztah 

Sorry for not finding this PR before leaving the comment
